### ISSUE TITLE
Error instantiating SqsClient - Unable to load an HTTP implementation from any provider in the chain. You must declare a dependency on an appropriate HTTP implementation or pass in an SdkHttpClient explicitly to the client builder.

### ISSUE THREAD

**Issue by rvojjala** [aws/aws-sdk-java-v2#1827]
Hi,

I am trying to instantiate SqsClient in my Scala project. When I build the image and deploy it in the EKS cluster, I keep getting the following error:

_software.amazon.awssdk.core.exception.SdkClientException: Unable to load an HTTP implementation from any provider in the chain. You must declare a dependency on an appropriate HTTP implementation or pass in an SdkHttpClient explicitly to the client builder._

I am testing my code in EKS cluster where Pods are configured with ServiceAccount. I verified the two AWS_ROLE_ARN and AWS_WEB_IDENTITY_TOKEN_FILE environment variables are populated with valid values. Expecting the authentication to take place using the web identity token. I tried various version of AWS SDK from 2.10.11 onwards. Currently using the 2.13.9.

I am trying to instantiate SqsClient as shown below:
```scala
         SqsClient.builder()
               .region(Region.US_EAST_1)
               .httpClient(ApacheHttpClient.builder().build())
              .build()
```
Going by some suggestions in the other forums,  I tried to instantiate WebIdentityTokenFileCredentialsProvider explicitly:
```scala
         SqsClient.builder()
                      .region(Region.US_EAST_1)
                      .credentialsProvider(WebIdentityTokenFileCredentialsProvider.create())
                      .build()
```
Out sbt config:

```sbt
libraryDependencies += "ch.qos.logback" % "logback-classic" % "1.2.3"
libraryDependencies += "com.typesafe.scala-logging" %% "scala-logging" % "3.9.2"
libraryDependencies += "com.typesafe" % "config" % "1.3.3"
libraryDependencies += "software.amazon.awssdk" % "sqs" % "2.13.9"
//libraryDependencies += "software.amazon.awssdk" % "sts" % "2.13.9"
```

I have also tried to use the following in sbt which includes a libraries for various services:
```sbt
libraryDependencies += "software.amazon.awssdk" % "aws-sdk-java" % "2.13.9"
```

Following is the exception stacktrace:

```
software.amazon.awssdk.core.exception.SdkClientException: Unable to load an HTTP implementation from any provider in the chain. You must declare a dependency on an appropriate HTTP implementation or pass in an SdkHttpClient explicitly to the client builder.
	at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build(SdkClientException.java:98)
	at software.amazon.awssdk.core.internal.http.loader.DefaultSdkHttpClientBuilder.lambda$buildWithDefaults$1(DefaultSdkHttpClientBuilder.java:49)
	at java.util.Optional.orElseThrow(Optional.java:290)
	at software.amazon.awssdk.core.internal.http.loader.DefaultSdkHttpClientBuilder.buildWithDefaults(DefaultSdkHttpClientBuilder.java:43)
	at software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder.lambda$resolveSyncHttpClient$5(SdkDefaultClientBuilder.java:269)
	at java.util.Optional.orElseGet(Optional.java:267)
	at software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder.resolveSyncHttpClient(SdkDefaultClientBuilder.java:269)
	at software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder.finalizeSyncConfiguration(SdkDefaultClientBuilder.java:220)
	at software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder.syncClientConfiguration(SdkDefaultClientBuilder.java:153)
	at software.amazon.awssdk.services.sqs.DefaultSqsClientBuilder.buildClient(DefaultSqsClientBuilder.java:27)
	at software.amazon.awssdk.services.sqs.DefaultSqsClientBuilder.buildClient(DefaultSqsClientBuilder.java:22)
	at software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder.build(SdkDefaultClientBuilder.java:124)
```

## Environment
* AWS Java SDK version used: 2.13.9
* JDK version used: 1.8
* Scala : 2.12.10
* Kubernetes: 1.15
* Operating System and version: Amazon Linux 2

**Comment by zoewangg** [comment]
Hi @rvojjala, to make sure I understand your question, you are still seeing issues with the following code snippet?

```java
         SqsClient.builder()
                      .region(Region.US_EAST_1)
                      .credentialsProvider(WebIdentityTokenFileCredentialsProvider.create())
                      .httpClient(ApacheHttpClient.builder().build())
                      .build()
```

As a side note, we recommend only importing individual service modules needed instead of the whole SDK, `aws-sdk-java`

**Comment by rvojjala** [comment]
Hi @zoewangg,

Yes. I still see issues with the code snippet that you mentioned. Also, we started with libraries for individual service modules only. But, in the process of debugging, included the whole SDK. But, surely we would like to switch back to the specific modules that we need once we have the issue resolved.

Thanks

**Comment by zoewangg** [comment]
I see. So explicitly setting the http client still doesn't work,  which means you probably don't have apache client in your classpath. Could you check if `ApacheSdkHttpService` is in the classpath? Could you also verify if `META-INF/services/software.amazon.awssdk.http.SdkHttpService` is in your jar?

Similar issue #446

**Comment by rvojjala** [comment]
Since I tested including the who SDK, I can see services/software.amazon.awssdk.http.SdkHttpService,
software.amazon.awssdk.http.apache.ApacheSdkHttpService,
software.amazon.awssdk.http.apache.ApacheHttpClient

and other classes in "software.amazon.awssdk.http" package are in the classpath.

**Comment by zoewangg** [comment]
Just to double check,  `apache-client` dependency is included in the final jar? Reading a bit more from https://github.com/aws/aws-sdk-java-v2/issues/446#issuecomment-395255114, seems like runtime dependencies are not in the jar. 

Could you also try explicitly declaring apache http client as compile dependency?

**Comment by rvojjala** [comment]
I tried with this as well. Unlike in Java, all the classes are unjarred and put into one common jar under their respective package structures. We use "sbt assembly" to build the jar file. 

Following is an example of how the jar looks:
com
 - mycompany
    - myorg
       - project
          <packages and classes>
software
 - amazon
    - awssdk
       - http
          - apache
          SdkHttpService.class (and other classes)
      - services
         - sqs
         - s3
      <other packages and classes>

**Comment by zoewangg** [comment]
Hmm, that's odd. Could you try setting the system property as follows?

```
 System.setProperty(SdkSystemSetting.SYNC_HTTP_SERVICE_IMPL.property(), "software.amazon.awssdk.http.apache.ApacheSdkHttpService");
```

**Comment by rvojjala** [comment]
Thank you so much. With this change, it worked for a few minutes and application was able to connect to SQS and processed messages.  But, after that it lost connectivity with SQS throwing the following . 

Going by your advice, I tried to play with the following setting, to see if it helps. But no luck.
```scala
System.setProperty(SdkSystemSetting.ASYNC_HTTP_SERVICE_IMPL.property, "software.amazon.awssdk.http.nio.netty.NettySdkAsyncHttpService")
```

```scala
02:38:41 DEBUG request {software.amazon.awssdk.utils.Logger debug} - Retryable error detected. Will retry in 71ms. Request attempt number 2
software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: Server failed to send complete response
	at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build(SdkClientException.java:98)
	at software.amazon.awssdk.core.exception.SdkClientException.create(SdkClientException.java:43)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.utils.RetryableStageHelper.setLastException(RetryableStageHelper.java:199)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.maybeRetryExecute(AsyncRetryableStage.java:143)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.lambda$attemptExecute$1(AsyncRetryableStage.java:125)
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760)
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage$WrappedErrorForwardingResponseHandler.onError(MakeAsyncHttpRequestStage.java:122)
	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.lambda$notifyIfResponseNotCompleted$2(ResponseHandler.java:399)
	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.runAndLogError(ResponseHandler.java:179)
	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.notifyIfResponseNotCompleted(ResponseHandler.java:399)
	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.channelInactive(ResponseHandler.java:151)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:260)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:246)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:239)
	at io.netty.handler.logging.LoggingHandler.channelInactive(LoggingHandler.java:197)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:260)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:246)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:239)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:81)
	at io.netty.handler.timeout.IdleStateHandler.channelInactive(IdleStateHandler.java:277)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:260)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:246)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:239)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelInactive(CombinedChannelDuplexHandler.java:418)
	at io.netty.handler.codec.ByteToMessageDecoder.channelInputClosed(ByteToMessageDecoder.java:386)
	at io.netty.handler.codec.ByteToMessageDecoder.channelInactive(ByteToMessageDecoder.java:351)
	at io.netty.handler.codec.http.HttpClientCodec$Decoder.channelInactive(HttpClientCodec.java:288)
	at io.netty.channel.CombinedChannelDuplexHandler.channelInactive(CombinedChannelDuplexHandler.java:221)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:260)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:246)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:239)
	at io.netty.handler.codec.ByteToMessageDecoder.channelInputClosed(ByteToMessageDecoder.java:386)
	at io.netty.handler.codec.ByteToMessageDecoder.channelInactive(ByteToMessageDecoder.java:351)
	at io.netty.handler.ssl.SslHandler.channelInactive(SslHandler.java:1084)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:260)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:246)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:239)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:81)
	at io.netty.handler.timeout.IdleStateHandler.channelInactive(IdleStateHandler.java:277)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:260)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:246)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:239)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelInactive(DefaultChannelPipeline.java:1405)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:260)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:246)
	at io.netty.channel.DefaultChannelPipeline.fireChannelInactive(DefaultChannelPipeline.java:901)
	at io.netty.channel.AbstractChannel$AbstractUnsafe$8.run(AbstractChannel.java:818)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: Server failed to send complete response
	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.notifyIfResponseNotCompleted(ResponseHandler.java:398)
```

**Comment by zoewangg** [comment]
Glad to hear it started to work. 

As regarding to `server faield to send complete response` error, it occurs when the connection is closed before the SDK was able to receive the response. Did you see this error with both sync client and async?

Did you have customized http client configuration such as `connectionMaxIdleTime`? Could you try the setting in https://github.com/aws/aws-sdk-java-v2/issues/1800#issuecomment-622507093?

**Comment by rvojjala** [comment]
Thank you. I will surely try that too. After adding the change you suggested, the issues were seen with Async clients. I tried to use Akka http client as we us Akka. After that, I did not see any errors. Also, removed passing http client object to SQSClient and S3Client.

We are able to see the code working. But, I will try the configuration that you suggested as well. 
Appreciate your help.

**Comment by debora-ito** [comment]
@rvojjala is there anything further pending from the SDK team on this issue?

**Comment by debora-ito** [comment]
Closing this, feel free to reopen if there's any further question.

**Comment by brownjeff** [comment]
@rvojjala I am guessing you solved this since it is 4 years later.  But I just ran into the same/similar issue.

In my case, the `META-INF/services` directory was being discarded due to the Merge Strategy setup.

To fix this, I used:

```
mainClass in assembly := Some("my.org.path.bogus.MainClass")

// Assembly settings
assemblyMergeStrategy in assembly := {
  case PathList("META-INF", "services", xs @ _*) => MergeStrategy.concat
  case PathList("META-INF", xs @ _*) => MergeStrategy.discard
  case "application.conf" => MergeStrategy.concat
  case _ => MergeStrategy.first
}
```

This ensures that service provider lists are concatenated, allowing all service implementations to be included in your fat JAR.

**_What is going on here?_**
The META-INF/services directory within a JAR file is a part of the Java Service Provider Interface (SPI) mechanism, which is used to enable service providers to be loaded and instantiated dynamically. This mechanism allows for modular service provider implementations that can be discovered and loaded at runtime without being hard-coded.

**_Contents of the META-INF/services Directory_**
In the META-INF/services directory, you will typically find one or more files named after the fully qualified interface or abstract class name they provide an implementation for. Each of these files contains the fully qualified names of the classes that implement the specified service, one per line. These classes must have a no-argument constructor so they can be instantiated by the service loader.

**_How It Works_**
When a service loader is used to load a particular service type, it reads the file in the META-INF/services directory that corresponds to the interface or abstract class of the service. The service loader then instantiates each listed implementation class using its no-argument constructor.

**_Common Use Cases_**
Here are a few common examples of where the SPI mechanism is used in Java libraries:

1. **Java SQL API (java.sql.Driver):** JDBC drivers register themselves as service providers, allowing them to be discovered and loaded without manual registration.

2. **Logging Frameworks (like SLF4J):** Different logging implementations can be swapped out by changing which service provider is included.

3. **JSON or XML Processing Libraries:** Libraries that process JSON or XML might allow for different parsing or processing implementations to be used interchangeably.

4. **HTTP Client Providers:** As mentioned in your earlier question, HTTP client libraries like those in Java 11 or other third-party libraries might use this mechanism to allow different HTTP client implementations to be used without code changes.
